### DISCUSSION TITLE
Apply resources using template error with "Kind and name are both required but at least one of them is missing from the manifest"

### DISCUSSION THREAD

**Discussion by ccfishk** [argoproj/argo-workflows#7700] [Q&A] [answered]
The error message is as following thrown by v3.2.7
message: Kind and name are both required but at least one of them is missing
          from the manifest
        reason: Error

Using template as following

`prereqs:
      template: |
        apiVersion: argoproj.io/v1alpha1
        kind: Workflow
        spec:
          entrypoint: entry
          serviceAccountName: addon-manager-workflow-installer-sa
          templates:
          - name: entry
            steps:
            - - name: prereq-resources
                template: submit
          - name: submit
            resource:
              action: apply
              manifest: |
                apiVersion: v1
                kind: Namespace
                metadata:
                  name: "{{workflow.parameters.namespace}}"
                ---`

I am aware there is a cherry-pick fix about this well-known back in April. 
Anything wrong with the template ? Or, this is another regression ?

**Comment by terrytangyuan** [comment, marked as answer]
You are missing metadata.name in your workflow manifest.

The fix last year that you are referring to is https://github.com/argoproj/argo-workflows/pull/5723. That is to remove group name as a requirement. Name and kind are still required for resource creation.

**Comment by ccfishk** [comment]
Do you mean `

metadata:
  generateName: workflow-a-name-`

**Reply by terrytangyuan** [reply]
Yes either name or generateName.

**Comment by ccfishk** [comment]
Am I insert in the right place, should be the name of resource argoproj.io/v1alpha1 ? right ?
still get the same complaining

      `apiVersion: argoproj.io/v1alpha1
        kind: Workflow
        metadata:
          generateName: event-router-prereqs-
        spec:
          entrypoint: entry
          serviceAccountName: addon-manager-workflow-installer-sa
          templates:
          - name: entry
            steps:
            - - name: prereq-resources
                template: submit

          - name: submit
            resource:
              action: apply
              manifest: |
                apiVersion: v1
                kind: Namespace
                metadata:
                  name: "{{workflow.parameters.namespace}}"
               `

**Reply by terrytangyuan** [reply]
Where did you define `workflow.parameters.namespace`?

BTW, please use the same thread. Otherwise it's hard to track the conversations.

**Reply by ccfishk** [reply]
`
    apiVersion: addonmgr.keikoproj.io/v1alpha1
    kind: Addon
    metadata:
      name: event-router
      namespace: addon-manager-system
    spec:
      pkgName: event-router
      pkgVersion: v0.2
      pkgType: composite
      pkgDescription: "Event router"
      params:
        namespace: addon-event-router-ns
        context:
          clusterName: "knock-knock"
          clusterRegion: us-west-2
      lifecycle:
        prereqs:
          template: |
            apiVersion: argoproj.io/v1alpha1
            kind: Workflow
            metadata:
              generateName: eventrouter-prereqs-
              labels:
                workflows.argoproj.io/controller-instanceid: addon-manager-workflow-controller
            spec:
              entrypoint: entry
              serviceAccountName: addon-manager-workflow-installer-sa
              templates:
              - name: entry
                steps:
                - - name: prereq-resources
                    template: submit
    
              - name: submit
                resource:
                  action: apply
                  manifest: |
                    apiVersion: v1
                    kind: Namespace
                    metadata:
                      name: "{{workflow.parameters.namespace}}"
`

**Reply by terrytangyuan** [reply]
Your `workflow.parameters.namespace` isn't defined anywhere but you are using it in your `submit` resource manifest. 

`"{{workflow.parameters.namespace}}"` is probably an empty string, which is why you are getting that error essage.

**Reply by ccfishk** [reply]
Sorry making your confused. But "{{workflow.parameters.namespace}}" is well defined though not pasted, and for debug, I even purposely replace the variables with static values.

The data below is pretty suspicious during container initialization stage, why metadata is empty ?
`name: ARGO_CONTAINER_NAME value: main - name: ARGO_TEMPLATE value: '{"name":"submit","inputs":{},"outputs":{},"metadata":{},"resource":{"action":"apply","manifest":"apiVersion: '`

**Reply by ccfishk** [reply]
You have a valid point - the error is thrown by resource.go, so it must be associated to resource definition.

**Reply by ccfishk** [reply]
atm, I sort believe it is a regression after applying my following test data that isolate the issue, trying to print out what exact resource this complaining is about

`apiVersion: addonmgr.keikoproj.io/v1alpha1
kind: Addon
metadata:
  name: event-router
  namespace: addon-manager-system
spec:
  pkgName: event-router
  pkgVersion: v0.2
  pkgType: composite
  pkgDescription: "Event router"
  params:
    namespace: addon-event-router-ns
    context:
      clusterName: "knock-knock"
      clusterRegion: us-west-2
  lifecycle:
    prereqs:
      template: |
        apiVersion: argoproj.io/v1alpha1
        kind: Workflow
        metadata:
          name: er-prereqs-
        spec:
          entrypoint: entry
          serviceAccountName: addon-manager-workflow-installer-sa
          templates:
          - name: entry
            steps:
            - - name: prereq-resources
                template: submit

          - name: submit
            resource:
              action: apply
              manifest: |
                apiVersion: v1
                kind: Namespace
                metadata:
                  name: addon-event-route-ns
`

**Reply by ccfishk** [reply]
I read the "resource" field from the wf pod which should be input of the executor/resource.go, none of them is lacking kind or name, takes a while to format the data

`value: '{"name":"submit","inputs":{},"outputs":{},"metadata":{},"resource":{"action":"apply","manifest":"apiVersion:
        v1\nkind: Namespace\nmetadata:\n    annotations: {}\n    labels:\n        app.kubernetes.io/managed-by:
        addonmgr.keikoproj.io\n        app.kubernetes.io/name: event-router\n        app.kubernetes.io/part-of:
        event-router\n        app.kubernetes.io/version: v0.2\n    name: addon-event-route-ns\n---\napiVersion:
        v1\nkind: ServiceAccount\nmetadata:\n    annotations: {}\n    labels:\n        app.kubernetes.io/managed-by:
        addonmgr.keikoproj.io\n        app.kubernetes.io/name: event-router\n        app.kubernetes.io/part-of:
        event-router\n        app.kubernetes.io/version: v0.2\n    name: event-router-sa\n    namespace:
        addon-event-route-ns\n---\napiVersion: v1\ndata:\n    config.json: |-\n        {\n          \"sink\":
        \"stdout\"\n        }\nkind: ConfigMap\nmetadata:\n    annotations: {}\n    labels:\n        app.kubernetes.io/managed-by:
        addonmgr.keikoproj.io\n        app.kubernetes.io/name: event-router\n        app.kubernetes.io/part-of:
        event-router\n        app.kubernetes.io/version: v0.2\n    name: event-router-cm\n    namespace:
        addon-event-route-ns\n---\napiVersion: rbac.authorization.k8s.io/v1beta1\nkind:
        ClusterRole\nmetadata:\n    annotations: {}\n    labels:\n        app.kubernetes.io/managed-by:
        addonmgr.keikoproj.io\n        app.kubernetes.io/name: event-router\n        app.kubernetes.io/part-of:
        event-router\n        app.kubernetes.io/version: v0.2\n    name: event-router-cr\nrules:\n    -
        apiGroups:\n        - \"\"\n      resources:\n        - events\n      verbs:\n        -
        get\n        - watch\n        - list\n---\napiVersion: rbac.authorization.k8s.io/v1beta1\nkind:
        ClusterRoleBinding\nmetadata:\n    annotations: {}\n    labels:\n        app.kubernetes.io/managed-by:
        addonmgr.keikoproj.io\n        app.kubernetes.io/name: event-router\n        app.kubernetes.io/part-of:
        event-router\n        app.kubernetes.io/version: v0.2\n    name: event-router-crb\nroleRef:\n    apiGroup:
        rbac.authorization.k8s.io\n    kind: ClusterRole\n    name: event-router-cr\nsubjects:\n    -
        kind: ServiceAccount\n      name: event-router-sa\n      namespace: addon-event-route-ns\n"}}'`

**Reply by ccfishk** [reply]
guess those "\n" might be the trouble maker

**Reply by ccfishk** [reply]
To isolate the issue, I only deploy one resource using prereqs template, and the WF pod initialization had following data 
`- name: ARGO_TEMPLATE
      value: '{"name":"submit","inputs":{},"outputs":{},"metadata":{},"resource":{"action":"apply","manifest":"apiVersion:
        v1\nkind: Namespace\nmetadata:\n  name: addon-event-route-ns\n---\napiVersion:
        v1\nkind: ServiceAccount\nmetadata:\n  name: event-router-sa\n  namespace:
        addon-event-route-ns\n"}}'
`
The wf-ctl does not fail on https://github.com/argoproj/argo-workflows/blob/master/workflow/executor/resource.go#L37, means the manifest is valid, however, fails on the checking https://github.com/argoproj/argo-workflows/blob/master/workflow/executor/resource.go#L67

which reads fields after unmarshal the string. What i gues is the "\n" between metadata and the name under it failed read the value. 

The above same template syntax works on wf-controller and argoexec v3.0.7, wf Pod is able to initialize and complete.

any suggestion here ?

**Comment by ccfishk** [comment]
Believe I am applying the correct syntax, however, ARGO wf controller does not recognize my metadata data, maybe other syntax trick 

`    - name: ARGO_CONTAINER_NAME
      value: main
    - name: ARGO_TEMPLATE
      value: '{"name":"submit","inputs":{},"outputs":{},"metadata":{},"resource":{"action":"apply","manifest":"apiVersion:
       '
`

`state:
      terminated:
        containerID: containerd://2bed9f8191452474dcf08b5edfe5a03a6b21e4bb498586d52b6c92fa940d6b50
        exitCode: 1
        finishedAt: "2022-01-31T04:20:27Z"
        message: Kind and name are both required but at least one of them is missing
          from the manifest
        reason: Error
        startedAt: "2022-01-31T04:20:26Z"

`

**Comment by ccfishk** [comment]
In the end, I come to conclude is a bug, posted all the details within issue

https://github.com/argoproj/argo-workflows/issues/7721
